File: javascript/test/client.test.ts
/** Client construction, headers, URL building and retry behaviour. */

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

import {
  APIConnectionError,
  APITimeoutError,
  DEFAULT_BASE_URL,
  KYCCentral,
  NotFoundError,
} from '../src/index.js';
import { BASE_URL, jsonResponse, mockFetch } from './helpers.js';

describe('construction', () => {
  const originalKey = process.env.KYCCENTRAL_API_KEY;
  const originalBase = process.env.KYCCENTRAL_BASE_URL;

  beforeEach(() => {
    delete process.env.KYCCENTRAL_API_KEY;
    delete process.env.KYCCENTRAL_BASE_URL;
  });

  afterEach(() => {
    if (originalKey === undefined) delete process.env.KYCCENTRAL_API_KEY;
    else process.env.KYCCENTRAL_API_KEY = originalKey;
    if (originalBase === undefined) delete process.env.KYCCENTRAL_BASE_URL;
    else process.env.KYCCENTRAL_BASE_URL = originalBase;
  });

  it('defaults to the production base URL and anonymous access', () => {
    const client = new KYCCentral();
    expect(client.baseUrl).toBe(DEFAULT_BASE_URL);
    expect(client.isAuthenticated).toBe(false);
  });

  it('reads credentials from the environment', () => {
    process.env.KYCCENTRAL_API_KEY = 'env-key';
    process.env.KYCCENTRAL_BASE_URL = 'https://dev-api.test.invalid/';
    const client = new KYCCentral();
    expect(client.isAuthenticated).toBe(true);
    // Trailing slash is normalised away so path joins never double up.
    expect(client.baseUrl).toBe('https://dev-api.test.invalid');
  });

  it('prefers explicit options over the environment', () => {
    process.env.KYCCENTRAL_API_KEY = 'env-key';
    const client = new KYCCentral({ apiKey: 'explicit', baseUrl: BASE_URL });
    expect(client.baseUrl).toBe(BASE_URL);
  });

  it.each([
    ['timeoutMs', { timeoutMs: 0 }],
    ['negative timeoutMs', { timeoutMs: -1 }],
    ['maxRetries', { maxRetries: -1 }],
  ])('rejects invalid %s', (_label, options) => {
    expect(() => new KYCCentral(options)).toThrow();
  });
});

describe('requests', () => {
  it('sends the API key and an Accept header', async () => {
    const fetch = mockFetch([jsonResponse([])]);
    const client = new KYCCentral({ apiKey: 'test-key', baseUrl: BASE_URL, fetch });

    await client.ruleSets.list();

    const [, init] = fetch.mock.calls[0]!;
    const headers = init!.headers as Record<string, string>;
    expect(headers['X-API-Key']).toBe('test-key');
    expect(headers['Accept']).toBe('application/json');
  });

  it('omits the API key header when anonymous', async () => {
    const fetch = mockFetch([jsonResponse({})]);
    const client = new KYCCentral({ baseUrl: BASE_URL, fetch });

    await client.jurisdictions.list();

    const headers = fetch.mock.calls[0]![1]!.headers as Record<string, string>;
    expect(headers['X-API-Key']).toBeUndefined();
  });

  it('sets Content-Type only when there is a body', async () => {
    const fetch = mockFetch([jsonResponse({}), jsonResponse({})]);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch });

    await client.sanctions.status();
    expect(
      (fetch.mock.calls[0]![1]!.headers as Record<string, string>)['Content-Type'],
    ).toBeUndefined();

    await client.sanctions.screenNames(['Alice Example']);
    expect((fetch.mock.calls[1]![1]!.headers as Record<string, string>)['Content-Type']).toBe(
      'application/json',
    );
  });

  it('versions business endpoints but not /health', async () => {
    const fetch = mockFetch([jsonResponse({ status: 'ok' }), jsonResponse([])]);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch });

    await client.health();
    expect(fetch.mock.calls[0]![0]).toBe(`${BASE_URL}/health`);

    await client.ruleSets.list();
    expect(fetch.mock.calls[1]![0]).toBe(`${BASE_URL}/v1/rule-sets`);
  });

  it('drops unset query parameters', async () => {
    const fetch = mockFetch([jsonResponse({ items: [] })]);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch });

    await client.companies.search('tesco');

    const url = new URL(fetch.mock.calls[0]![0]);
    expect(url.searchParams.get('q')).toBe('tesco');
    expect(url.searchParams.has('items_per_page')).toBe(false);
    expect(url.searchParams.has('start_index')).toBe(false);
  });

  it('repeats array parameters once per value', async () => {
    const fetch = mockFetch([jsonResponse({ company_number: '00445790', flags: [] })]);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch });

    await client.kyc.assess('00445790', {
      confirmedMediaUrls: ['https://a.example', 'https://b.example'],
    });

    const url = new URL(fetch.mock.calls[0]![0]);
    expect(url.searchParams.getAll('confirmed_media_url')).toEqual([
      'https://a.example',
      'https://b.example',
    ]);
  });

  it('percent-encodes path segments', async () => {
    const fetch = mockFetch([jsonResponse({})]);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch });

    // Officer ids are opaque upstream tokens and can contain URL-significant bytes.
    await client.companies.officerAppointments('abc/def+gh');

    expect(fetch.mock.calls[0]![0]).toContain('abc%2Fdef%2Bgh');
  });

  it.each(['', '   '])('rejects blank path segments (%j)', async (blank) => {
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch: mockFetch([]) });
    await expect(client.companies.get(blank)).rejects.toThrow(/companyNumber/);
  });

  it('sends custom default headers', async () => {
    const fetch = mockFetch([jsonResponse([])]);
    const client = new KYCCentral({
      apiKey: 'k',
      baseUrl: BASE_URL,
      fetch,
      defaultHeaders: { 'X-Trace': 'abc' },
    });

    await client.ruleSets.list();

    expect((fetch.mock.calls[0]![1]!.headers as Record<string, string>)['X-Trace']).toBe('abc');
  });
});

describe('retries', () => {
  beforeEach(() => {
    vi.useFakeTimers({ shouldAdvanceTime: true });
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  it('retries retryable statuses then succeeds', async () => {
    const fetch = mockFetch([
      jsonResponse({ detail: 'warming up' }, 503),
      jsonResponse([{ id: 'default' }]),
    ]);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch, maxRetries: 2 });

    await expect(client.ruleSets.list()).resolves.toEqual([{ id: 'default' }]);
    expect(fetch).toHaveBeenCalledTimes(2);
  });

  it('does not retry client errors', async () => {
    const fetch = mockFetch([jsonResponse({ detail: 'Company not found' }, 404)]);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch, maxRetries: 3 });

    await expect(client.companies.get('99999999')).rejects.toBeInstanceOf(NotFoundError);
    expect(fetch).toHaveBeenCalledTimes(1);
  });

  it('surfaces connection errors after exhausting retries', async () => {
    const fetch = vi.fn().mockRejectedValue(new Error('ECONNREFUSED'));
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch, maxRetries: 1 });

    await expect(client.ruleSets.list()).rejects.toBeInstanceOf(APIConnectionError);
    expect(fetch).toHaveBeenCalledTimes(2);
  });

  it('reports timeouts distinctly from other connection failures', async () => {
    const timeout = new Error('The operation was aborted due to timeout');
    timeout.name = 'TimeoutError';
    const fetch = vi.fn().mockRejectedValue(timeout);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch, maxRetries: 0 });

    await expect(client.ruleSets.list()).rejects.toBeInstanceOf(APITimeoutError);
  });

  it('does not retry a caller-initiated abort', async () => {
    const controller = new AbortController();
    const fetch = vi.fn().mockImplementation(() => {
      controller.abort();
      const error = new Error('aborted');
      error.name = 'AbortError';
      return Promise.reject(error);
    });
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch, maxRetries: 3 });

    await expect(client.ruleSets.list({ signal: controller.signal })).rejects.toBeInstanceOf(
      APITimeoutError,
    );
    // An intentional cancellation must not be retried into three more requests.
    expect(fetch).toHaveBeenCalledTimes(1);
  });
});

File: elixir/test/kyccentral/client_test.exs
defmodule KYCCentral.ClientTest do
  @moduledoc "Client construction, headers, URL building and retry behaviour."

  use ExUnit.Case, async: true

  alias KYCCentral.{Error, Stub}

  describe "new/1" do
    test "defaults to the production base URL and anonymous access" do
      client = KYCCentral.new(api_key: nil)

      assert client.base_url == "https://api.kyccentral.co.uk"
      refute KYCCentral.authenticated?(client)
    end

    test "reads credentials from the environment" do
      System.put_env("KYCCENTRAL_API_KEY", "env-key")
      System.put_env("KYCCENTRAL_BASE_URL", "https://dev-api.test.invalid/")
      on_exit(fn -> System.delete_env("KYCCENTRAL_API_KEY") end)
      on_exit(fn -> System.delete_env("KYCCENTRAL_BASE_URL") end)

      client = KYCCentral.new()

      assert KYCCentral.authenticated?(client)
      # Trailing slash is normalised away so path joins never double up.
      assert client.base_url == "https://dev-api.test.invalid"
    end

    test "prefers explicit options over the environment" do
      System.put_env("KYCCENTRAL_API_KEY", "env-key")
      on_exit(fn -> System.delete_env("KYCCENTRAL_API_KEY") end)

      client = KYCCentral.new(api_key: "explicit", base_url: Stub.base_url())

      assert client.api_key == "explicit"
      assert client.base_url == Stub.base_url()
    end

    test "treats a blank key as anonymous" do
      refute KYCCentral.authenticated?(KYCCentral.new(api_key: ""))
    end

    test "rejects nonsense settings" do
      assert_raise ArgumentError, fn -> KYCCentral.new(receive_timeout: 0) end
      assert_raise ArgumentError, fn -> KYCCentral.new(max_retries: -1) end
    end
  end

  describe "requests" do
    test "sends the API key, accept and user-agent headers" do
      {client, agent} = Stub.client([Stub.json([])])

      assert {:ok, []} = KYCCentral.RuleSets.list(client)

      assert Stub.header(agent, "x-api-key") == "test-key"
      assert Stub.header(agent, "accept") == "application/json"
      assert Stub.header(agent, "user-agent") =~ "kyccentral-elixir/"
    end

    test "omits the API key header when anonymous" do
      {client, agent} = Stub.client([Stub.json(%{})], api_key: nil)

      assert {:ok, _} = KYCCentral.Jurisdictions.list(client)
      assert Stub.header(agent, "x-api-key") == nil
    end

    test "sets content-type only when there is a body" do
      {client, agent} = Stub.client([Stub.json(%{}), Stub.json(%{})])

      assert {:ok, _} = KYCCentral.Sanctions.status(client)
      assert Stub.header(agent, "content-type") == nil

      assert {:ok, _} = KYCCentral.Sanctions.screen_names(client, ["Alice Example"])
      assert Stub.header(agent, "content-type", 1) == "application/json"
    end

    test "sends custom default headers" do
      {client, agent} = Stub.client([Stub.json([])], headers: [{"x-trace", "abc"}])

      assert {:ok, _} = KYCCentral.RuleSets.list(client)
      assert Stub.header(agent, "x-trace") == "abc"
    end

    test "versions business endpoints but not health" do
      {client, agent} = Stub.client([Stub.json(%{"status" => "ok"}), Stub.json([])])

      assert {:ok, %{"status" => "ok"}} = KYCCentral.health(client)
      assert Stub.path(agent) == "/health"

      assert {:ok, _} = KYCCentral.RuleSets.list(client)
      assert Stub.path(agent, 1) == "/v1/rule-sets"
    end

    test "drops unset query parameters" do
      {client, agent} = Stub.client([Stub.json(%{"items" => []})])

      assert {:ok, _} = KYCCentral.Companies.search(client, "tesco")

      assert Stub.query_value(agent, "q") == "tesco"
      # Neither optional parameter was supplied, so neither should appear.
      assert Stub.query_value(agent, "items_per_page") == nil
      assert Stub.query_value(agent, "start_index") == nil
    end

    test "repeats list parameters once per value" do
      {client, agent} = Stub.client([Stub.json(Stub.assessment_payload())])

      assert {:ok, _} =
               KYCCentral.KYC.assess(client, "00445790",
                 confirmed_media_urls: ["https://a.example", "https://b.example"]
               )

      assert Stub.query_values(agent, "confirmed_media_url") == [
               "https://a.example",
               "https://b.example"
             ]
    end

    test "percent-encodes path segments" do
      {client, agent} = Stub.client([Stub.json(%{})])

      # Officer ids are opaque upstream tokens and can contain URL-significant bytes.
      assert {:ok, _} = KYCCentral.Companies.officer_appointments(client, "abc/def+gh")
      assert Stub.call(agent).url =~ "abc%2Fdef%2Bgh"
    end

    test "rejects blank path segments before making a request" do
      {client, agent} = Stub.client([Stub.json(%{})])

      assert {:error, %Error{kind: :invalid_argument, message: message}} =
               KYCCentral.Companies.get(client, "   ")

      assert message =~ "company_number"
      assert Stub.call_count(agent) == 0
    end
  end

  describe "retries" do
    test "retries a retryable status and then succeeds" do
      {client, agent} =
        Stub.client(
          [Stub.json(%{"detail" => "warming up"}, 503), Stub.json([%{"id" => "default"}])],
          max_retries: 2
        )

      assert {:ok, [%{"id" => "default"}]} = KYCCentral.RuleSets.list(client)
      assert Stub.call_count(agent) == 2
    end

    test "does not retry client errors" do
      {client, agent} =
        Stub.client([Stub.json(%{"detail" => "Company not found"}, 404)], max_retries: 3)

      assert {:error, %Error{kind: :not_found}} = KYCCentral.Companies.get(client, "99999999")
      # A 404 will not become true on a second attempt.
      assert Stub.call_count(agent) == 1
    end

    test "gives up on a retryable status once retries are exhausted" do
      {client, agent} = Stub.client([Stub.json(%{"detail" => "down"}, 503)], max_retries: 1)

      assert {:error, %Error{kind: :service_unavailable}} = KYCCentral.RuleSets.list(client)
      assert Stub.call_count(agent) == 2
    end

    test "surfaces connection failures after retrying" do
      {client, agent} =
        Stub.failing_client({:failed_connect, [{:to_address, {~c"host", 443}}]}, max_retries: 1)

      assert {:error, %Error{kind: :connection}} = KYCCentral.RuleSets.list(client)
      assert Stub.call_count(agent) == 2
    end

    test "reports timeouts distinctly from other connection failures" do
      {client, _agent} = Stub.failing_client(:timeout, max_retries: 0)

      assert {:error, %Error{kind: :timeout, message: message}} = KYCCentral.RuleSets.list(client)
      assert message =~ "timed out"
    end

    test "recognises a nested httpc connect timeout" do
      reason = {:failed_connect, [{:to_address, {~c"host", 443}}, {:inet, [:inet], :timeout}]}
      {client, _agent} = Stub.failing_client(reason, max_retries: 0)

      assert {:error, %Error{kind: :timeout}} = KYCCentral.RuleSets.list(client)
    end
  end
end
